core/src/main/java/org/infinispan/persistence/factory/CacheStoreFactoryRegistry.java

```diff
@@ -20,7 +21,7 @@
 
    private static final Log log = LogFactory.getLog(CacheStoreFactoryRegistry.class);
 
-   private List<CacheStoreFactory> factories = new ArrayList<>();
+   private List<CacheStoreFactory> factories = Collections.synchronizedList(new ArrayList<>());
```

Comment:
The root cause was lack of synchronization ;)
Since it is used only during startup - I think synchronizing it using simplest possible way will be a good fit.